### PR TITLE
Adding Panda environments

### PR DESCRIPTION
**Pull request by qgallouedec**
## Description

I propose to add the environments included in the `panda-gym` package to `rl-baselines3-zoo`. For more details, see Issue #123.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*). **(Not sure about this one...)**
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: hyperparams/her.yml [hunk 1]
```diff
@@ -92,6 +92,97 @@ FetchReach-v1:
   online_sampling: True
   normalize: True
 
+# Panda Robotic Env
```
Comment (araffin) [inline hyperparams/her.yml:95]: Please only update the TQC/SAC params as HER is now a replay bufffer and the hyperparams here are only there for legacy reasons.

**Review by araffin — COMMENTED** [review]
One minor thing missing: could you please add the evaluation files generated by the benchmark script? (You will need to do `git add -f` on them are they are ignored by default)
If you don't have them anymore, I will regenerate them ;)

**Comment by araffin** [comment]
>One minor thing missing: could you please add the evaluation files generated by the benchmark script? (You will need to do git add -f on them are they are ignored by default)
If you don't have them anymore, I will regenerate them ;)

And this would be for TQC+HER using latest SB3 version (I merged the wrong agents in the rl trained agent PR)

**Comment by qgallouedec** [comment]
> One minor thing missing: could you please add the evaluation files generated by the benchmark script? (You will need to do git add -f on them are they are ignored by default)
> If you don't have them anymore, I will regenerate them ;)

Done 👍 

> And this would be for TQC+HER using latest SB3 version (I merged the wrong agents in the rl trained agent PR)

I opened a new PR to fix this.

**Comment by araffin** [comment]
>I opened a new PR to fix this.

Thanks.
Only the hyperparameters for TQC+HER on Pandas envs are missing now (btw, why not use the same as for the Fetch envs?).

I wanted to activate logging the success rate for the Pandas envs but as you used master branch, I have no rights to push to it...
You can see my changes in https://github.com/DLR-RM/rl-baselines3-zoo/tree/qgallouedec/master

**Comment by qgallouedec** [comment]
> Only the hyperparameters for TQC+HER on Pandas envs are missing now (btw, why not use the same as for the Fetch envs?).

Here they are, I forgot them in the last commit.

> I wanted to activate logging the success rate for the Pandas envs but as you used master branch, I have no rights to push to it...
> You can see my changes in [ff58f66](https://github.com/DLR-RM/rl-baselines3-zoo/commit/ff58f6698e7c402141393b3837787e0668cdfbfa) on the branch: https://github.com/DLR-RM/rl-baselines3-zoo/tree/qgallouedec/master

Is there anything I can do to solve this problem?

**Comment by qgallouedec** [comment]
Actually, hyperparameters are similar as the ones used for the Fetch env

**Comment by araffin** [comment]
>In fact, the hyperparameters are identical to those used for the Fetch env.

hmm, that would ideal but the config of the trained agents from your PR does not match the one from the yaml file ...

https://github.com/DLR-RM/rl-trained-agents/pull/7/files

**Comment by qgallouedec** [comment]
> hmm, that would ideal but the config of the trained agents from your PR does not match the one from the yaml file ...

I just checked, it seems to me that they match. What difference am I missing?

**Comment by araffin** [comment]
for instance for https://github.com/DLR-RM/rl-trained-agents/pull/7/files#diff-cdb52017ca9ec28f74f88bca54337b152ff71f4a37e8d2505c5546d36c2a1090 (PandaPickAndPlace-v1/config.yml ), the batch size is 512 and not 2048, no?

**Comment by qgallouedec** [comment]
Actually, the agents were trained with the same hyperparameters as those used for the Fetch environments:

This was not the case when I submitted [PR#5](https://github.com/DLR-RM/rl-trained-agents/pull/5) (in [PandaPush-v1 config file for example](https://github.com/DLR-RM/rl-trained-agents/blob/master/her/PandaPush-v1_1/PandaPush-v1/config.yml), the batch size was indeed 2048 and not 512, for example). However, since it was necessary to change the agents because they were trained with an old version of `rl-baselines3-zoo` (see your [comment](https://github.com/DLR-RM/rl-baselines3-zoo/pull/124#issuecomment-868377300)), I took the opportunity to change the config file in https://github.com/DLR-RM/rl-baselines3-zoo/pull/124, in order that the hyperparameters of Fetch and Panda are identical (see https://github.com/qgallouedec/rl-baselines3-zoo/commit/89f8455a78ac687aa4f1b85cb691f9bc6e188c0c).  Then, I deleted the old agents in https://github.com/DLR-RM/rl-trained-agents/pull/7/commits/7c56cbcf4ead15dcb6d3a9cb63bcb2fca72ec2b7 and submitted the new ones in https://github.com/DLR-RM/rl-trained-agents/pull/7/commits/79f863f4fb7934051a8f751352a8426931471566.

**Comment by araffin** [comment]
>Actually, the agents were trained with the same hyperparameters as those used for the Fetch environments:

oh I see, my bad, it seems that I commited hyperparameters that were not the one I wanted ^^"

**Comment by araffin** [comment]
but does it work as well as with the previous ones? (on Fetch envs, it was ok luckily)

**Comment by qgallouedec** [comment]
> but does it work as well as with the previous ones? (on Fetch envs, it was ok luckily)

I'm not able to quantify it yet, but I suspect that the previous hyparameters are working worse, especially for slide. I plan to look into this in detail.

**Comment by araffin** [comment]
>I'm not able to quantify it yet, but I suspect that the previous hyparameters are working worse, especially for slide. I plan to look into this in detail.

I've got the training graphs (using the zoo plot script):

Before:
Training_Episodic_Reward


after:
Training_Episodic_Reward_3

I'll try to re-train the agents and update the PR, no worry ;)

**Comment by qgallouedec** [comment]
> I'll try to re-train the agents and update the PR, no worry ;)

Unfortunately, this confirms my intuition ;)

**Comment by araffin** [comment]
While re-training on Panda envs with the new hyperparameters, I noticed something weird: if you use more than one evaluation env `--n-eval-envs 2` (for instance), then the evaluation will be far from the real performance...

The problem can be solved by using sub-process vec envs (`--vec-env subproc`), so I think there may be some issues with shared memory...
Also, it seems that seeding does not work on Panda envs (I cannot have reproducible results).

**Comment by qgallouedec** [comment]
Thanks for the feedback, I'll fix it, I'll keep you informed.

**Comment by qgallouedec** [comment]
Hi, I made an update of `panda-gym` (version v1.1.0, see [pypi.org](https://pypi.org/project/panda-gym/)) which solves the two problems you pointed out (seeding and compatibility with vectorized env). 
I've tested with the command lines in your comment, now we get **exactly** the same results :)

```bash
$ python train.py --algo tqc --env PandaReach-v1 --eval-freq 5000 --num-threads 2 --n-eval-envs 4 --eval-episodes 20 --seed 9676371 -n 5000
[...]
Eval num_timesteps=5000, episode_reward=-9.00 +/- 16.00
Episode length: 50.00 +/- 0.00
Success rate: 85.00%
New best mean reward!
---------------------------------
| eval/              |          |
|    mean_ep_length  | 50       |
|    mean_reward     | -9       |
|    success_rate    | 0.85     |
| rollout/           |          |
|    ep_len_mean     | 50       |
|    ep_rew_mean     | -43.5    |
|    success rate    | 0.14     |
| time/              |          |
|    episodes        | 100      |
|    fps             | 142      |
|    time_elapsed    | 35       |
|    total timesteps | 5000     |
| train/             |          |
|    actor_loss      | -2.46    |
|    critic_loss     | 0.00143  |
|    ent_coef        | 0.0234   |
|    ent_coef_loss   | -10.3    |
|    learning_rate   | 0.001    |
|    n_updates       | 3999     |
---------------------------------
$ python train.py --algo tqc --env PandaReach-v1 --eval-freq 5000 --num-threads 2 --n-eval-envs 1 --eval-episodes 20 --seed 9676371 -n 5000
[...]
Eval num_timesteps=5000, episode_reward=-4.35 +/- 5.73
Episode length: 50.00 +/- 0.00
Success rate: 100.00%
New best mean reward!
---------------------------------
| eval/              |          |
|    mean_ep_length  | 50       |
|    mean_reward     | -4.35    |
|    success_rate    | 1.0      |
| rollout/           |          |
|    ep_len_mean     | 50       |
|    ep_rew_mean     | -43.5    |
|    success rate    | 0.14     |
| time/              |          |
|    episodes        | 100      |
|    fps             | 138      |
|    time_elapsed    | 35       |
|    total timesteps | 5000     |
| train/             |          |
|    actor_loss      | -2.46    |
|    critic_loss     | 0.00143  |
|    ent_coef        | 0.0234   |
|    ent_coef_loss   | -10.3    |
|    learning_rate   | 0.001    |
|    n_updates       | 3999     |
---------------------------------

```

**Comment by araffin** [comment]
Perfect, thanks =)

**Comment by araffin** [comment]
to simplify everything, I've open another PR that I will merge ;) https://github.com/DLR-RM/rl-baselines3-zoo/pull/130

next time, please use a new branch so I have the right to push to it ;) (master branch of your fork is protected by default)

**Comment by qgallouedec** [comment]
> next time, please use a new branch so I have the right to push to it ;) (master branch of your fork is protected by default)

Ok, I didn't know, thanks :)

**Pull request closed without merging**